src/rgw/rgw_rest_pubsub.cc

```diff
@@ -77,43 +104,91 @@ std::optional<rgw::IAM::Policy> get_policy_from_text(req_state* const s,
     ldout(s->cct, 1) << "failed to parse policy: '" << policy_text
                      << "' with error: " << e.what() << dendl;
     s->err.message = e.what();
-    return std::nullopt;
+    return boost::none;
   }
 }
 
-int verify_topic_owner_or_policy(req_state* const s,
-                                 const rgw_pubsub_topic& topic,
-                                 const std::string& zonegroup_name,
-                                 const uint64_t op) {
-  if (s->auth.identity->is_owner_of(topic.owner)) {
-    return 0;
+using rgw::IAM::Effect;
+using rgw::IAM::Policy;
+
+bool verify_topic_permission(const DoutPrefixProvider* dpp, req_state* s,
+                             const rgw_owner& owner, const rgw::ARN& arn,
+                             const boost::optional<Policy>& policy,
+                             uint64_t op)
+{
+  if (s->auth.identity->get_account()) {
+    const bool account_root = (s->auth.identity->get_identity_type() == TYPE_ROOT);
+    if (!s->auth.identity->is_owner_of(owner)) {
+      ldpp_dout(dpp, 4) << "cross-account request for resource owner "
+          << owner << " != " << s->owner.id << dendl;
+      // cross-account requests evaluate the identity-based policies separately
+      // from the resource-based policies and require Allow from both
+      const auto identity_res = evaluate_iam_policies(
+          dpp, s->env, *s->auth.identity, account_root, op, arn,
+          {}, s->iam_identity_policies, s->session_policies);
+      if (identity_res == Effect::Deny) {
+        return false;
+      }
+      const auto resource_res = evaluate_iam_policies(
+          dpp, s->env, *s->auth.identity, false, op, arn,
+          policy, {}, {});
+      return identity_res == Effect::Allow && resource_res == Effect::Allow;
+    } else {
+      // require an Allow from either identity- or resource-based policy
+      return Effect::Allow == evaluate_iam_policies(
+          dpp, s->env, *s->auth.identity, account_root, op, arn,
+          policy, s->iam_identity_policies, s->session_policies);
+    }
+  }
+
+  constexpr bool account_root = false;
```

Comment:
i was thinking this function in all cases (whether an account, if cross-account or a general user (non-account) users, we are always checking for all 3 policies
- s->iam_identity_policies
-  s->session_policies
- policy (resource policy == topic policy)

then why not just simplify the function and have a single call
`  const auto effect = evaluate_iam_policies(
      dpp, s->env, *s->auth.identity, account_root, op, arn,
      policy, s->iam_identity_policies, s->session_policies);`
rather so many if else ??

Reply:
the cross-account case has to make two separate calls, one with only the resource policy and another with only the identity policies

you're right that the same-account and non-account cases are similar enough that they might be combined, but they'd still need a branch after that for the handling of Effect::Pass: that's treated as an implicy deny for account users vs. an implicit allow for non-account users (depending on ownership check, rgw_topic_require_publish_policy, etc)

personally, i prefer the current form where the account logic is completely separated from the non-account logic. i followed that same pattern with verify_bucket_permission() etc in https://github.com/ceph/ceph/pull/54333

Reply:
so why does the cross-account case has to make two separate calls, one with only the resource policy and another with only the identity policies ??? and calls cannot be combine ?

Reply:
this `evaluate_iam_policies()` function implements the logic for [Determining whether a request is allowed or denied within an account](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html#policy-eval-denyallow). as long as none of the policies match a Deny statement, then an Allow from either identity- or resource-based policy is good enough to grant access

[Determining whether a cross-account request is allowed](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic-cross-account.html#policy-eval-cross-account) describes the algorithm for cross-account policy evaluation:
> For cross-account requests, the requester in the trusted AccountA must have an identity-based policy. That policy must allow them to make a request to the resource in the trusting AccountB. Additionally, the resource-based policy in AccountB must allow the requester in AccountA to access the resource.
>
>When you make a cross-account request, AWS performs two evaluations. AWS evaluates the request in the trusting account and the trusted account. For more information about how a request is evaluated within a single account, see [Determining whether a request is allowed or denied within an account](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html#policy-eval-denyallow). The request is allowed only if both evaluations return a decision of Allow.

here the two policy types (resource vs. identity) are managed by different accounts. these rules mean that cross-account access is only granted when both accounts agree. the two separate calls to `evaluate_iam_policies()` are how we require both evaluations to return Allow

Reply:
while identity policy can restrict access based on Resource, they often use wildcards. take the [AmazonS3FullAccess](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonS3FullAccess.html) managed policy, for example:
```json
    {
      "Effect" : "Allow",
      "Action" : [
        "s3:*",
        "s3-object-lambda:*"
      ],
      "Resource" : "*"
    }
```
if not for those special cross-account evaluation rules, identity policies like this could grant blanket access to all resources

Reply:
ohh ok i see now, the variable `account_root` is a difference between 2 calls as the for identity its true but for the resource it is false as we are validating the resource_policy for different account and hence pass false.

Reply:
right. also note that the first call omits the resource policy:
```c++
          dpp, s->env, *s->auth.identity, account_root, op, arn,
          {}, s->iam_identity_policies, s->session_policies);
```
while the second omits the identity/session policies:
```c++
          dpp, s->env, *s->auth.identity, false, op, arn,
          policy, {}, {});
```

Reply:
sorry @cbodley i had another doubt on cross account users workflow. I remember you saying in the meeting the as per AWS, topics cannot be shared between cross account users (unlike bucket), even with resource_policy giving access to createtopic or listtopic or deletetopic.
however based on above condition and statement we are still evaluating the identity policy n resource policy for cross account users and if both pass then its allowed sharing ? so we allow cross account sharing?
as per the comment `     // cross-account requests evaluate the identity-based policies separately
      // from the resource-based policies and require Allow from both` 
`

Reply:
this cross-account logic will only apply to sns:Publish permissions for PutBucketNotificationConfiguration, not for the Topic apis which can only view the topic namespace of the requesting user's account

Reply:
got it, i see that `verify_topic_permission` is called explicitly only for CreateNotification, for all other calls `verify_user_permission` is triggered first and if topic exist (in case of create n delete) `verify_topic_permission` is called, but then for cross-account get_topic will return ENOENT so it will invoke `verify_user_permission `.

Unfortunately this works as expected, but from code readability this is not great.
would have been much simpler all s3/sns calls would call `verify_topic_permission` and then inside that detect if its cross-account and then deny or validate all there.
someone new reading code will get confused between `verify_topic_permission` and `verify_user_permission`

Reply:
> got it, i see for all calls `verify_user_permission` is triggered first and if topic exist (in case of create n delete) `verify_topic_permission` is called, but then for cross-account get_topic will return ENOENT so it will invoke `verify_user_permission `.

sorry, i don't quite follow. to which op are you referring?

> someone new reading code will get confused between verify_topic_permission and verify_user_permission

for s3, we have separate calls for verify_user_permission(), verify_bucket_permission(), and verify_object_permission() for the different levels of the api. i think that's a good model in general, but i can see how it's confusing to mix the levels in a single op. i tried to provide useful code comments where that happens in the CreateTopic/DeleteTopic ops. do you have any suggestions to clarify those?

Reply:
how about a generic `verify_permission` tht is called by all topic/notification api's and then based on account-users call verify_user_permission from inside that else call the `verify-topic-permission`. 
this was we do not have to worry about different logic for CreateTopic/DeleteTopci ?

Reply:
would that really be an improvement? if it's confusing for CreateTopic and DeleteTopic to handle the optional-topic case, moving that into a generic function means that it would add confusion for all other ops too